### ISSUE TITLE
现代Python数据科学实战：从数据处理到机器学习

### ISSUE THREAD

**Issue by sbumsid24** [sbumsid24/DJ#3]
Python已成为数据科学领域的事实标准语言，凭借其丰富的生态系统和易用性，在数据处理、分析和机器学习等方面占据主导地位。本文将深入探讨Python在数据科学中的应用，涵盖NumPy、Pandas、Matplotlib和Scikit-learn等核心库，并提供大量实用代码示例，帮助读者掌握数据科学工作流程。

1. 数据处理基础
1.1 NumPy数组操作
import numpy as np

# 创建数组
arr1 = np.array([1, 2, 3, 4, 5])
arr2 = np.arange(0, 10, 2)  # 0到10，步长2
arr3 = np.random.rand(5)  # 5个随机数

print("数组1:", arr1)
print("数组2:", arr2)
print("随机数组:", arr3)

# 数组运算
print("\n数组运算:")
print("加法:", arr1 + arr2[:len(arr1)])
print("乘法:", arr1 * 2)
print("点积:", np.dot(arr1, arr1))

# 矩阵操作
matrix = np.array([[1, 2], [3, 4]])
print("\n矩阵:")
print(matrix)
print("矩阵转置:")
print(matrix.T)
print("矩阵求逆:")
print(np.linalg.inv(matrix))
1.2 Pandas数据处理
import pandas as pd

# 创建DataFrame
data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 40],
    'Salary': [50000, 60000, 70000, 80000]
}
df = pd.DataFrame(data)

print("原始数据:")
print(df)

# 数据操作
print("\n数据操作:")
print("平均年龄:", df['Age'].mean())
print("最高薪水:", df['Salary'].max())
print("年龄大于30的人:")
print(df[df['Age'] > 30])

# 添加新列
df['Bonus'] = df['Salary'] * 0.1
print("\n添加奖金列:")
print(df)

# 分组统计
print("\n按年龄分组统计:")
print(df.groupby('Age').mean())
2. 数据可视化
2.1 Matplotlib基础图表
import matplotlib.pyplot as plt

# 折线图
x = np.linspace(0, 10, 100)
y = np.sin(x)

plt.figure(figsize=(10, 5))
plt.plot(x, y, label='sin(x)', color='blue')
plt.title('正弦函数曲线')
plt.xlabel('X轴')
plt.ylabel('Y轴')
plt.legend()
plt.grid(True)
plt.show()

# 柱状图
categories = ['A', 'B', 'C', 'D']
values = [15, 25, 30, 20]

plt.figure(figsize=(8, 4))
plt.bar(categories, values, color=['red', 'green', 'blue', 'yellow'])
plt.title('分类数据柱状图')
plt.xlabel('类别')
plt.ylabel('数值')
plt.show()

# 散点图
np.random.seed(42)
x = np.random.rand(50)
y = x + np.random.normal(0, 0.1, 50)

plt.figure(figsize=(8, 6))
plt.scatter(x, y, alpha=0.6)
plt.title('随机数据散点图')
plt.xlabel('X值')
plt.ylabel('Y值')
plt.show()
2.2 Seaborn高级可视化
import seaborn as sns

# 加载示例数据集
tips = sns.load_dataset('tips')

# 箱线图
plt.figure(figsize=(8, 6))
sns.boxplot(x='day', y='total_bill', data=tips)
plt.title('每日账单分布')
plt.show()

# 热力图
corr = tips.corr()
plt.figure(figsize=(8, 6))
sns.heatmap(corr, annot=True, cmap='coolwarm')
plt.title('相关性热力图')
plt.show()

# 配对图
sns.pairplot(tips, hue='sex')
plt.suptitle('数据特征关系图', y=1.02)
plt.show()
3. 机器学习实战
3.1 数据预处理
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split

# 加载数据
from sklearn.datasets import load_iris
iris = load_iris()
X, y = iris.data, iris.target

# 数据分割
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

# 特征标准化
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("\n标准化前5个样本:")
print(X_train[:5])
print("\n标准化后5个样本:")
print(X_train_scaled[:5])
3.2 分类模型训练
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

# 训练模型
model = LogisticRegression(max_iter=200)
model.fit(X_train_scaled, y_train)

# 预测评估
y_pred = model.predict(X_test_scaled)

print("\n分类报告:")
print(classification_report(y_test, y_pred))

print("\n混淆矩阵:")
print(confusion_matrix(y_test, y_pred))
3.3 随机森林与调参
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# 定义参数网格
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5, 10]
}

# 网格搜索
rf = RandomForestClassifier(random_state=42)
grid_search = GridSearchCV(rf, param_grid, cv=5, n_jobs=-1)
grid_search.fit(X_train_scaled, y_train)

# 最佳模型
best_rf = grid_search.best_estimator_
print("\n最佳参数:", grid_search.best_params_)
print("最佳分数:", grid_search.best_score_)

# 评估最佳模型
y_pred_rf = best_rf.predict(X_test_scaled)
print("\n随机森林分类报告:")
print(classification_report(y_test, y_pred_rf))
4. 深度学习入门
4.1 TensorFlow神经网络
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# 构建模型
model = Sequential([
    Dense(64, activation='relu', input_shape=(4,)),
    Dense(32, activation='relu'),
    Dense(3, activation='softmax')
])

# 编译模型
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# 训练模型
history = model.fit(X_train_scaled, y_train, 
                    epochs=50, batch_size=16,
                    validation_split=0.2, verbose=0)

# 评估模型
loss, accuracy = model.evaluate(X_test_scaled, y_test)
print(f"\n测试集准确率: {accuracy:.2f}")

# 绘制训练曲线
plt.figure(figsize=(10, 5))
plt.plot(history.history['accuracy'], label='训练准确率')
plt.plot(history.history['val_accuracy'], label='验证准确率')
plt.title('模型训练过程')
plt.xlabel('Epoch')
plt.ylabel('准确率')
plt.legend()
plt.show()
4.2 PyTorch实现CNN
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms

# 数据转换
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5,), (0.5,))
])

# 加载MNIST数据
train_set = datasets.MNIST(root='./data', train=True, download=True, transform=transform)
test_set = datasets.MNIST(root='./data', train=False, download=True, transform=transform)

# 定义CNN模型
class CNN(nn.Module):
    def __init__(self):
        super(CNN, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)
    
    def forward(self, x):
        x = torch.relu(self.conv1(x))
        x = torch.max_pool2d(x, 2)
        x = torch.relu(self.conv2(x))
        x = torch.max_pool2d(x, 2)
        x = torch.flatten(x, 1)
        x = torch.rel
https://hunyinzx.com/baike/2268.html
https://hunyinzx.com/baike/2269.html
https://hunyinzx.com/baike/2270.html
https://hunyinzx.com/baike/2271.html
https://hunyinzx.com/baike/2272.html
https://suanmingw.cn/baike/2607.html
https://suanmingw.cn/baike/2608.html
https://suanmingw.cn/baike/2609.html
https://suanmingw.cn/baike/2610.html
https://suanmingw.cn/baike/2612.html
https://suanmingw.cn/baike/2611.html
https://stcjzx.com/baike/3089.html
https://stcjzx.com/baike/3090.html
https://stcjzx.com/baike/3091.html
https://stcjzx.com/baike/3092.html
https://stcjzx.com/baike/3093.html
https://stcjzx.com/baike/3094.html
https://hunyinzx.com/baike/2323.html
